homeassistant/components/tplink_eap/device_tracker.py

```diff
@@ -0,0 +1,135 @@
+"""Tracker for Wifi clients connected to a TP-Link EAP."""
+import logging
+from typing import List, Set
+
+import async_timeout
+from pytleap import Client, Eap, PytleapError
+
+from homeassistant.components.device_tracker import (
+    ATTR_MAC,
+    DOMAIN as DEVICE_TRACKER_DOMAIN,
+    SOURCE_TYPE_ROUTER,
+)
+from homeassistant.components.device_tracker.config_entry import ScannerEntity
+from homeassistant.components.device_tracker.const import SCAN_INTERVAL
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import entity_registry
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the device_tracker by config_entry."""
+    coordinator = TpLinkEapCoordinator(
+        hass, config_entry, async_add_entities, hass.data[DOMAIN][config_entry.entry_id]
+    )
+
+    # Initialize already tracked entities
+    tracked: Set[str] = set()
+    registry = await entity_registry.async_get_registry(hass)
+    known_entities: List[Entity] = []
+    for entity in registry.entities.values():
+        if (
+            entity.domain == DEVICE_TRACKER_DOMAIN
+            and entity.config_entry_id == config_entry.entry_id
+        ):
+            tracked.add(entity.unique_id)
+            known_entities.append(WifiClientEntity(coordinator, entity.unique_id))
+    async_add_entities(known_entities, True)
+    coordinator.tracked_entity_macs = tracked
+
+    if not known_entities:
+        # If no entities exist, need to bootstrap process
```

Comment:
I'm not 100% sure why I need to do that here.
If I remove the test protect, then it looks like `coordinator.async_refresh()` is called twice.
If I remove the call to `coordinator.async_refresh()` altogether, then when there are no already defined entities (eg. after setting up the integration), then the coordinator appears to not be updated...

Reply:
The coordinator will only refresh automatically when there are entities that have connected callbacks to the coordinator. So if we need to update without entities, to be able to find entities, we may need to make sure there's at least one callback. We could eg make a callback for the entity addition instead of building that into the coordinator update function.

Here's the interface to add a refresh callback to the coordinator:
https://github.com/home-assistant/core/blob/1bb535aa6784fcb1af69749ec8eea2a1207456b9/homeassistant/helpers/update_coordinator.py#L77